### PR TITLE
add collective op based DistributedOptimizer design

### PR DESCRIPTION
**Pull request by guru4elephant**

### CODE REVIEW

File: python/paddle/fluid/incubate/fleet/collective/__init__.py [hunk 1]
```diff
@@ -48,8 +48,25 @@ def stop_worker(self):
...
+                    FullPrecisionOptimizer(optmizer, strategy)
+        if strategy.local_sgd:
+            if strategy.use_fp16:
+                # should have an independent optimizer inherited from
+                # CollectiveOpBasedOptimizer
+                pass
+            else:
+                # should have an independent optimizer inherited from
+                # CollectiveOpBasedOptimizer
+                pass
```
Comment (gavin1332) [inline python/paddle/fluid/incubate/fleet/collective/__init__.py]: could we use decorator pattern to replace these ifelse? e.g. if we require algorithm chain as momentum + fp16 + localsgd, we declare
`opt = LocalSGD(MixedPrecision(Momentum(learning_rate=learning_rate, momentum=0.9)))`

File: python/paddle/fluid/incubate/fleet/collective/__init__.py [hunk 2]
```diff
@@ -69,6 +84,78 @@ def save_persistables(self, executor, dirname, main_program=None):
 fleet = Collective()
 
 
+class CollectiveOpBasedOptimizer(DistributedOptimizer):
```
Comment (gavin1332) [inline python/paddle/fluid/incubate/fleet/collective/__init__.py:128]: the similar names, CollectiveOpBasedOptmizer and CollectiveOptimizer, make me confused a bit, as CollectiveOptimizer is original, could we change CollectiveOpBasedOptmizer to some other? As u r the NAMING MASTER, do u have any ideas?

File: python/paddle/fluid/incubate/fleet/collective/__init__.py (hunk 1 above)
Comment (guru4elephant) [reply python/paddle/fluid/incubate/fleet/collective/__init__.py]: a factory or decorator can be design as free. Generally, a user only need to define the strategy to do distributed optimization, and instance of DistributedOptimizer is determined by strategy a user defined.

File: python/paddle/fluid/incubate/fleet/collective/__init__.py (hunk 2 above)
Comment (guru4elephant) [reply python/paddle/fluid/incubate/fleet/collective/__init__.py:128]: I generally want to make CollectiveOpBasedOptmizer as a base class for all DistributedOptimizers that use collective op to do optimization. Maybe CollectiveOpOptimizerBase is better.

**Review by gavin1332 — APPROVED** [review]

**Pull request merged**